### PR TITLE
Updates data processing logic to remove dependency on hardcoded chat templates

### PR DESCRIPTION
**Pull request by RobotSail**
In order to use the training library today, we need to manually define the chat template and rely on specifically parsing the special tokens.

This introduces a number of issues for consumers where other models cannot easily be used without some inital effort.

This PR resolves this issue by introducing a new form of data processing which is able to arbitrarily process data in the messages format and apply the appropriate unmasking policy.

Specifically, the new data processing script now reads the `unmask` field to determine whether to unmask all messages (aside from the system message) or just the assistant responses (skills training).

This PR adds:
- A new `process_data` API into the `data_process` module which is meant to be the main backwards-compatible user-friendly entrypoint for data processing
- A new `process_messages_into_input_ids` API in `instructlab.training.data_process` which supports the conversion of any arbitrary HF messages dataset into a masked format that the training library can consume
- Conversion logic to provide compatibility with legacy datasets that contain `pretraining` message roles (based on the format found here: https://github.com/instructlab/sdg/blob/3ca69b6395e8a9cae62f5edaf002f3240a308b66/src/instructlab/sdg/datamixing.py#L380) 
- Updates the training loop to load in a model with the arbitrary tokenizer + chat template
- Example script that shows how someone can call the new data processing APIs

### CODE REVIEW

File: src/instructlab/training/data_process.py [hunk 1]
```diff
@@ -1,20 +1,32 @@
...
+from instructlab.training.types import Message, ProcessedMessages
 from instructlab.training.utils import log_rank_0, retrieve_chat_template, setup_logger
 
+# Constants
+MASK_TOKEN = "<|MASK|>"
+UNMASK_BEGIN_TOKEN = "<|UNMASK_BEGIN|>"
+UNMASK_END_TOKEN = "<|UNMASK_END|>"
+
+
+disable_caching()
```
Comment (Maxusmusti) [inline src/instructlab/training/data_process.py:28]: NOTE: remove

File: src/instructlab/training/utils.py [hunk 2]
```diff
@@ -201,6 +215,8 @@ def check_flash_attn_enabled(disable_flash_attn: bool, use_dolomite: bool) -> bo
 def make_collate_fn(
     pad_token_id, use_dolomite=False, flash_enabled=True, max_batch_len=60000
 ):
+    # some tokenizers may not have a pad token id, so we use -100 as a default
+    pad_token_id = pad_token_id if pad_token_id is not None else -100
```
Comment (Maxusmusti) [inline src/instructlab/training/utils.py:219]: NOTE: Remove old artifact

**Review by Maxusmusti — APPROVED** [review]
LGTM

**Review by aldopareja — APPROVED** [review]
